### PR TITLE
Minor fixes in IOException catch and metrics registry

### PR DESCRIPTION
**Pull request by jsjtzyy**
1. When log segment instantiation encounters IOException, no matter
freeing log segment succeeds or not, ensure that store exception is
correctly thrown in ensureCapacity method.
2. Register the custom percentile metrics during RouterMetrics
initialization rather than register them every time adaptive tracker is
created. This will avoid unnecessary overhead.

### CODE REVIEW

File: ambry-router/src/test/java/com.github.ambry.router/CryptoJobHandlerTest.java [hunk 1]
```diff
@@ -69,7 +69,7 @@ public CryptoJobHandlerTest() throws IOException, GeneralSecurityException {
     cryptoService = new GCMCryptoServiceFactory(verifiableProperties, REGISTRY).getCryptoService();
     cryptoJobHandler = new CryptoJobHandler(DEFAULT_THREAD_COUNT);
     referenceClusterMap = new MockClusterMap();
-    routerMetrics = new NonBlockingRouterMetrics(referenceClusterMap);
+    routerMetrics = new NonBlockingRouterMetrics(referenceClusterMap, null);
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/CryptoJobHandlerTest.java]: I purposely set  `RouterConfig`  = `null` when creating `NonBlockingRouterMetrics` in several tests. This is to verify that even if router config is mistakenly set to null in the ctor of `NonBlockingRouterMetrics`, there would be no failure during instantiation.

File: ambry-store/src/main/java/com.github.ambry.store/Log.java [hunk 2]
```diff
@@ -470,17 +466,15 @@ private void ensureCapacity(long writeSize) throws StoreException {
           new LogSegment(segmentNameAndFilename.getFirst(), newSegmentFile, segmentCapacity, metrics, true);
       segmentsByName.put(segmentNameAndFilename.getFirst(), newSegment);
     } catch (StoreException e) {
-      logger.error("Failed to create new log segment {} with store exception: ", segmentNameAndFilename.getFirst(), e);
       try {
```
Comment (cgtz) [inline ambry-store/src/main/java/com.github.ambry.store/Log.java]: a test for this block will be good, but its ok if its too cumbersome.

File: ambry-api/src/main/java/com.github.ambry/store/StoreException.java [hunk 3]
```diff
@@ -13,6 +13,9 @@
  */
 package com.github.ambry.store;
 
+import java.util.Objects;
+
+
 public class StoreException extends Exception {
   public static final String INTERNAL_ERROR_STR =
```
Comment (cgtz) [inline ambry-api/src/main/java/com.github.ambry/store/StoreException.java]: Can these be private or package-private

File: ambry-store/src/main/java/com.github.ambry.store/Log.java (hunk 2 above)
Comment (jsjtzyy) [reply ambry-store/src/main/java/com.github.ambry.store/Log.java]: test added

File: ambry-api/src/main/java/com.github.ambry/store/StoreException.java (hunk 3 above)
Comment (jsjtzyy) [reply ambry-api/src/main/java/com.github.ambry/store/StoreException.java]: changed it to package-private

**Comment by jsjtzyy** [comment]
./gradlew build && ./gradlew test succeeded. @cgtz @zzmao  reminder to review.

**Pull request merged**